src/diagonal.jl

```diff
@@ -6,35 +6,49 @@ function EmptyDiagonalPreconditioner(A::AbstractMatrix)
     return DiagonalPreconditioner{eltype(D), typeof(D)}(D)
 end
 function DiagonalPreconditioner(A::AbstractMatrix)
+    size(A, 1) == size(A, 2) || throw(ArgumentError("matrix must be square"))
     D = diag(A, 0)
+    # Since diag(::SparseMatrixCSC) return a SparseVector we convert it to a Vector since
+    # indexing is much faster for that case, and the diagonal should be dense anyway.
+    # The type check is added here instead of having dispatches on SparseMatrixCSC since
+    # that would not catch e.g. Hermitian{SparseMatrixCSC} and other variants.
+    if D isa SparseVector
+        if nnz(D) == size(A, 1)
+            # Fast path in case all values of the diagonal are stored (this should always be
+            # true since otherwise we will divide by 0 later)
+            D = D.nzval
+        else
+            # Fallback (TODO: Better to just error?)
+            D = convert(Vector, D)
+        end
+    end
     return DiagonalPreconditioner{eltype(D), typeof(D)}(D)
 end
 function diag!(D, A::AbstractMatrix)
-    length(D) == size(A,1) == size(A,2) || throw("D and A sizes are not compatible.")
-    @inbounds @simd for i in 1:length(D)
-        D[i] = A[i,i]
+    if !(axes(D, 1) == axes(A, 1) == axes(A, 2))
+        throw(ArgumentError("incompatible indices for input arguments"))
+    end
+    @inbounds for (i, j, k) in zip(eachindex(D), axes(A, 1), axes(A, 2))
+        D[i] = A[j, k]
     end
     return
 end
 function UpdatePreconditioner!(D::DiagonalPreconditioner, K::AbstractMatrix)
     diag!(D.D, K)
     return D
 end
-@inline ldiv!(C::DiagonalPreconditioner, b) = ldiv!(b, C, b)
-@inline function ldiv!(y, C::DiagonalPreconditioner, b)
-    @inbounds @simd for j ∈ 1:size(y, 2)
-        for i ∈ 1:length(C.D)
-            y[i,j] = b[i,j] / C.D[i]
-        end
+ldiv!(C::DiagonalPreconditioner, b) = ldiv!(b, C, b)
+function ldiv!(y, C::DiagonalPreconditioner, b)
+    if !(axes(y, 1) == axes(b, 1) == axes(C.D, 1) && axes(y, 2) == axes(b, 2))
+        throw(ArgumentError("incompatible indices for input arguments"))
     end
-    return y
-end
-@inline function (\)(C::DiagonalPreconditioner, b)
-    y = zero(b)
-    @inbounds @simd for j ∈ 1:size(y, 2)
-        for i ∈ 1:length(C.D)
-            y[i,j] = b[i,j] / C.D[i]
+    @inbounds for (yj, bj) in zip(axes(y, 2), axes(b, 2))
```

Comment:
Is this faster than using `ldiv!` for matrices? I know you improved the previous implementation but I am curious if we can do better.

Reply:
I am talking about something like this
```julia
ldiv!(rand(2,2), Diagonal(rand(2)), rand(2,2))
```

Reply:
I will merge this one and we can iterate on it in another PR if needed.

Reply:
I am not sure. This implementation only requires indexing to work so seems a bit easier, and `ldiv!` with a `Diagonal` probably does the same thing under the hood anyway.